arbor/mechcat.cpp

```diff
@@ -49,245 +51,510 @@
+
+        auto apply_globals = [this, &apply_deriv](auto& self, const std::string& name, mechanism_overrides& over) -> void {
+            if (auto p = value_by_key(derived_map_, name)) {
+                self(self, p->parent, over);
+                apply_deriv(over, *p);
+            }
+        };
+
+        util::optional<derivation> implicit_deriv;
+        if (!defined(name)) {
+            if (auto deriv = derive(name)) {
+                implicit_deriv = std::move(deriv.first());
+            }
+            else {
+                return deriv.second();
+            }
         }
-    };
-    apply_globals(apply_globals, name, mech.second);
-    return mech;
-}
 
-void mechanism_catalogue::register_impl(std::type_index tidx, const std::string& name, std::unique_ptr<mechanism> mech) {
-    const mechanism_info& info = (*this)[name];
+        apply_globals(apply_globals, implicit_deriv? implicit_deriv->parent: name, over);
+        if (implicit_deriv) {
+            apply_deriv(over, implicit_deriv.value());
+        }
 
-    if (mech->fingerprint()!=info.fingerprint) {
-        throw fingerprint_mismatch(name);
+        return over;
     }
 
-    impl_map_[name][tidx] = std::move(mech);
+    // Schemata for (un-derived) mechanisms.
+    string_map<mechanism_info_ptr> info_map_;
+
+    // Parent and global setting values for derived mechanisms.
+    string_map<derivation> derived_map_;
+
+    // Prototype register, keyed on mechanism name, then backend type (index).
+    string_map<std::unordered_map<std::type_index, mechanism_ptr>> impl_map_;
+};
+
+// Mechanism catalogue method implementations.
+
+mechanism_catalogue::mechanism_catalogue():
+    state_(new catalogue_state)
+{}
+
+mechanism_catalogue::mechanism_catalogue(mechanism_catalogue&& other) = default;
+mechanism_catalogue& mechanism_catalogue::operator=(mechanism_catalogue&& other) = default;
+
+mechanism_catalogue::mechanism_catalogue(const mechanism_catalogue& other):
+    state_(new catalogue_state(*other.state_))
+{}
+
+mechanism_catalogue& mechanism_catalogue::operator=(const mechanism_catalogue& other) {
+    state_.reset(new catalogue_state(*other.state_));
+    return *this;
 }
 
-void mechanism_catalogue::copy_impl(const mechanism_catalogue& other) {
-    info_map_.clear();
-    for (const auto& kv: other.info_map_) {
-        info_map_[kv.first] = make_unique<mechanism_info>(*kv.second);
+void mechanism_catalogue::add(const std::string& name, mechanism_info info) {
+    if (state_->defined(name)) {
+        throw duplicate_mechanism(name);
     }
+    state_->bind(name, std::move(info));
+}
 
-    derived_map_.clear();
-    for (const auto& kv: other.derived_map_) {
-        const derivation& v = kv.second;
-        derived_map_[kv.first] = {v.parent, v.globals, v.ion_remap, make_unique<mechanism_info>(*v.derived_info)};
-    }
+bool mechanism_catalogue::has(const std::string& name) const {
+    return state_->defined(name) || state_->derive(name);
```

Comment:
Yup, this could go either way. I was interpreting `has(name)` as being equivalent to: can I use `name` as a mechanism. Maybe `valid` would be a better name for the method?

Reply:
Okay. Naming it `valid` or adding a comment to `has` would make it clearer, I think.